### PR TITLE
PROD-757, adding router diagnostics

### PR DESCRIPTION
**Pull request by f1401martin**

### CODE REVIEW

File: nio/router/base.py [hunk 1]
```diff
@@ -320,6 +324,22 @@ def notify_signals(self, block, signals, output_id):
                             len(signals_to_send),
                             block.name(),
                             receiver_data.block.name()))
+
+                    # if diagnostics are enabled, notify
+                    # through management signal delivery
+                    if self._diagnostics:
+                        self.notify_management_signal(
```
Comment (mattdodge) [inline nio/router/base.py]: This means every time any block notifies a signal we also notify a management signal, right? That is not a scalable solution, we shouldn't double the amount of signals sent through pub/sub in the default implementation. We'll probably want to set up some kind of asynchronous notification type thing to do this instead.

File: nio/router/base.py [hunk 2]
```diff
@@ -149,6 +160,16 @@ def configure(self, context):
                         sender_block._default_output.id)
                     self._receivers[sender_block_name].extend(parsed_receivers)
 
+    def start(self):
+        super().start()
+        if self._diagnostic_manager:
+            self._diagnostic_manager.start()
```
Comment (mattdodge) [inline nio/router/base.py]: Since it is a `Runner` I think we should call `do_start()` instead of `start()`

File: nio/router/diagnostic.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        if self._job:
+            self._job.cancel()
+
+        # Send what we have saved
+        self._send_diagnostic()
+        super().stop()
+
+    def on_signal(self, signal):
+        with self._signals_lock:
+            self._signals.append(signal)
```
Comment (mattdodge) [inline nio/router/diagnostic.py]: I would actually rather have this diagnostic manager consolidate these signals together. In other words, if I've set up the diagnostic interval to notify every minute, I wouldn't want to still get a ton of signals sent about everything that happened in that minute. Given this is only diagnostic data we don't need that level of granularity. Instead, I'd rather have one signal sent for every output-input block combination with the number of signals sent between them during that minute.

File: nio/router/base.py [hunk 4]
```diff
@@ -94,12 +99,18 @@ def configure(self, context):
...
         self._mgmt_signal_handler = context.mgmt_signal_handler
+        self._instance_id = context.instance_id
         self._clone_signals = \
             context.settings.get("clone_signals", True)
         if self._clone_signals:
             self.logger.info('Set to clone signals for multiple receivers')
         self._check_signal_type = \
             context.settings.get("check_signal_type", True)
+        self._diagnostics = \
+            context.settings.get("diagnostics", True)
```
Comment (mattdodge) [inline nio/router/base.py]: I know we use it elsewhere, but how do these settings get properly converted to booleans when they are specified in files? I would think we would need to call a `getbool` on it at some point but don't know where that is happening

File: nio/router/diagnostic.py (hunk 3 above)
Comment (f1401martin) [reply nio/router/diagnostic.py]: good catch, makes good sense

File: nio/router/base.py (hunk 4 above)
Comment (f1401martin) [reply nio/router/base.py]: context.settings is a dictionary built in nio-core which is where the Settings.getboolean call happens

Comment (mattdodge) [reply nio/router/base.py]: Aha! I see that PR now, just merged it. Makes sense, thanks

**Pull request merged**